api/emoji.go

```diff
@@ -147,11 +150,45 @@ func uploadEmojiImage(id string, imageData *multipart.FileHeader) *model.AppErro
 	if config, _, err := image.DecodeConfig(bytes.NewReader(buf.Bytes())); err != nil {
 		return model.NewLocAppError("uploadEmojiImage", "api.emoji.upload.image.app_error", nil, err.Error())
 	} else if config.Width > MaxEmojiWidth || config.Height > MaxEmojiHeight {
-		return model.NewLocAppError("uploadEmojiImage", "api.emoji.upload.large_image.app_error", nil, "")
-	}
-
-	if err := WriteFile(buf.Bytes(), getEmojiImagePath(id)); err != nil {
-		return err
+		data := buf.Bytes()
+		newbuf := bytes.NewBuffer(nil)
+		if info, err := model.GetInfoForBytes(imageData.Filename, data); err != nil {
+			return err
+		} else if info.MimeType == "image/gif" {
+			if gif_data, err := gif.DecodeAll(bytes.NewReader(data)); err != nil {
+				return model.NewLocAppError("uploadEmojiImage", "api.emoji.upload.large_image.gif_decode_error", nil, "")
+			} else {
+				resized_gif := resizeEmojiGif(gif_data)
+				if err := gif.EncodeAll(newbuf, resized_gif); err != nil {
+					return model.NewLocAppError("uploadEmojiImage", "api.emoji.upload.large_image.gif_encode_error", nil, "")
+				}
+				if err := WriteFile(newbuf.Bytes(), getEmojiImagePath(id)); err != nil {
+					return err
+				}
+			}
+		} else {
+			if img, _, err := image.Decode(bytes.NewReader(data)); err != nil {
+				return model.NewLocAppError("uploadEmojiImage", "api.emoji.upload.large_image.decode_error", nil, "")
+			} else {
+				resized_image := resizeEmoji(img, config.Width, config.Height)
+				if info.MimeType == "image/jpeg" {
```

Comment:
I think you can just pick a consistent type here for all resized emoji instead of making it match the type of the original image. That'll simplify the code a bit since you'll only need one `Encode` call.